### ISSUE TITLE
New method to enable stubbing

### ISSUE THREAD

**Issue by richswrighttj** [ForbesLindesay/sync-request#50]
Hi

Love the library, I've got a valid use for it but....It's hard to test. It's fallen into the same syntax trap as 'request', namely the lack of a function to stub on. 

You can see people discussing it about request here - http://stackoverflow.com/questions/20050507/creating-request-stub-with-sinon-in-mocha

Request does have 'get' and 'post' to enable stubbing. Could we get the same on this library? It would be most helpful

**Comment by ForbesLindesay** [comment]
I think there are two approaches here that would make sense:

1. Use a library like mockery that will allow you to just mock the return value of `require` calls.  I haven't used mockery specifically, so can't vouch for it working, but I think this approach in general is sound as it allows you to mock whatever you want without having to damage the API of the modules you want to mock.
2. Add explicit `.mockRequest` and `.clearMocks` functions to this library.  This is really simple to implement and could have a consistent API client and server side.
3. Add `.record` and `.replay` functions to this library.  This has the advantage of allowing you to not manually write out the mocks for each function, but it would not work client side.

## mockRequest(method, url, options, response)

Any subsequent call to `request` that passes the same method (case insensitive), url and options (tested via [deep-equal](https://www.npmjs.com/package/deep-equal) will return the `response`.  The `response` object must match the API of http-response-object.  On the server side, the body **must** be a buffer, on the client side, it could be a string.  This can be verified by some asserts in `mockRequest`.

Once mockRequest has been called, any un-mocked requests should throw an exception, along with details of exactly what method, url and options were passed in.

## clearMocks()

Empties the list of currently mocked requests and restores the normal behaviour of actually making requests.

## record(dirname)

Makes all requests use a customised `cache` based on the [file-cache](https://github.com/ForbesLindesay/http-basic/blob/master/lib/file-cache.js) but with the [cache-utils](https://github.com/ForbesLindesay/http-basic/blob/master/lib/cache-utils.js) methods overridden so `isExpired` and `canCache` returns `true` and `isMatch` returns `false`.

## replay(dirname)

Makes all requests use a customised `cache` based on the [file-cache](https://github.com/ForbesLindesay/http-basic/blob/master/lib/file-cache.js) but with the [cache-utils](https://github.com/ForbesLindesay/http-basic/blob/master/lib/cache-utils.js) methods overridden so `isMatch` and `canCache` returns `true` and `isExpired` returns `false`.  Maybe make `setResponse` throw an exception.

**Comment by richswrighttj** [comment]
These all sound rather over blown. A simple method for each http verb or even a generic method for all the requests would suffice.

e.g. request.get('google.com') or request.makeRequest('GET', 'google.com);

I opened (and closed) a pull request on this last week. I didn't want to waste writing the code if you didn't feel like it fitted into the style of the project. If you will let it in, then I'm happy to write the code.

**Comment by ForbesLindesay** [comment]
That approach only lets you mock requests using those helper methods, not all requests.  I also don't think it's a good idea to make a worse API in order to enable simpler testing when you can already do mocking of the current API.

Probably the easiest way would be to use [babel-plugin-module-alias](https://www.npmjs.com/package/babel-plugin-module-alias) which you could then use to point `'sync-request'` at `'./mock-sync-request.js'`.

**Comment by RichardWright** [comment]
Exactly. You get to be really explicit with your mocks. Instead of everything going through the same setup, you get the extra readability from the method name AND less 'magic' strings being passed around.

Extra readability without any extra code? That's a win.

**Comment by ForbesLindesay** [comment]
I think the method reduces readability.  Having been using it for almost 2 years, I'm much happier without the methods.

Having `request('get', ...args)` behave differently to `request.get(...args)` would be ridiculously confusing/magical.  You should not be writing mocks that rely on details like that.

**Comment by RichardWright** [comment]
Did you read the code? It uses the main request method, it's just lost a parameter. They behave exactly the same.

**Comment by ForbesLindesay** [comment]
Yes I did read the code.  They don't behave the same if you mock one of them, which was after all the entire point of this request.

**Comment by RichardWright** [comment]
That really depends on how the test is setup and is in the hands of the user? Are you worried about people not realising they are making real requests?

**Comment by ForbesLindesay** [comment]
I'm worried that the OP intends to write a mock that goes:

```js
const request = require('sync-request');

request.get = function (path) {
  return myFakeResponse;
};
```

which leads to `request.get` and `request` then not behaving the same.  It also means that the same issue will come up for every single module that uses this single export style.  The fix already exists in that there are plenty of modules to allow you to mock the single export of another module.

**Comment by RichardWright** [comment]
That approach constrains the verbs you can work with surely? Although I don't see it as a large issue, different methods allow reusage of the request mock (eg, stubbing for a get then a post).

**Comment by jhenaoz** [comment]
i like the solution for make a method syncRequestMock, like angular with the HttpBackendObject, or also extend the support for make syncRequest.get, also with the older support syncRequest(httpMethod...other params.). angular is a good example on that. (Backwards compatibility with new features easy for test and mock)
